src/controlplane/scraper.go

```diff
@@ -0,0 +1,282 @@
+type Providers struct {
+	K8s kubernetes.Interface
+}
+
+// Scraper takes care of getting metrics all control plane instances based on the configuration.
+type Scraper struct {
+	Providers
+	logger               log.Logger
+	config               *config.Config
+	k8sVersion           *version.Info
+	components           []component
+	informerClosers      []chan<- struct{}
+	podListerByNamespace map[string]v1.PodNamespaceLister
+	inClusterConfig      *rest.Config
+}
+
+// ScraperOpt are options that can be used to configure the Scraper
+type ScraperOpt func(s *Scraper) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) ScraperOpt {
+	return func(s *Scraper) error {
+		if logger == nil {
+			return fmt.Errorf("logger canont be nil")
+		}
+
+		s.logger = logger
+
+		return nil
+	}
+}
+
+// WithRestConfig returns an OptionFunc to change the restConfig from default empty config.
+func WithRestConfig(restConfig *rest.Config) ScraperOpt {
+	return func(s *Scraper) error {
+		if restConfig == nil {
+			return fmt.Errorf("restConfig canont be nil")
+		}
+		s.inClusterConfig = restConfig
+		return nil
+	}
+}
+
+// Close will signal internal informers to stop running.
+func (s *Scraper) Close() {
+	for _, ch := range s.informerClosers {
+		close(ch)
+	}
+}
+
+// NewScraper initialize its internal informers and components.
+// After use, informers should be closed by calling Close().
+func NewScraper(config *config.Config, providers Providers, options ...ScraperOpt) (*Scraper, error) {
+	var err error
+	s := &Scraper{
+		config:    config,
+		Providers: providers,
+		// TODO: An empty implementation of the logger interface would be better
+		logger:               log.New(false, io.Discard),
+		podListerByNamespace: make(map[string]v1.PodNamespaceLister),
+		components:           newComponents(config.ControlPlane),
+		inClusterConfig:      &rest.Config{},
+	}
+
+	for i, opt := range options {
+		if err := opt(s); err != nil {
+			return nil, fmt.Errorf("applying config option #%d: %w", i, err)
+		}
+	}
+
+	// TODO If this could change without a restart of the pod we should run it each time we scrape data,
+	// possibly with a reasonable cache Es: NewCachedDiscoveryClientForConfig
+	s.k8sVersion, err = providers.K8s.Discovery().ServerVersion()
+	if err != nil {
+		return nil, fmt.Errorf("fetching K8s version: %w", err)
+	}
+
+	// Building pod lister and closers for pod autodisover.
+	s.buildLister()
+
+	return s, nil
+}
+
+// Run scraper collect the data populating the integration entities.
+func (s *Scraper) Run(i *integration.Integration) error {
+	var jobs []*scrape.Job
+
+	for _, component := range s.components {
+		var job *scrape.Job
+
+		var err error
+
+		// Static endpoint take precedence over autodisover and fails if external endpoint
+		// cannot be scraped.
+		if component.StaticEndpointConfig != nil {
+			s.logger.Debugf("Building %q external endpoint job")
```

Comment:
```suggestion
			s.logger.Infof("Using static endpoint for component %s")
```

⬆️  Argument for `%q` seems to be missing

Reply:
Uh, didn't see it was being printed on each iteration. Definitely should not be `Info` then, good catch!